### PR TITLE
[16.0][IMP] stock_picking_group_by_partner_by_carrier: Control the original_group_id propagation.

### PR DESCRIPTION
**Pull request by lmignon**

### CODE REVIEW

**Comment by lmignon** [comment]
cc @rousseldenis @jbaudoux

File: stock_picking_group_by_partner_by_carrier/models/stock_move.py [hunk 1]
```diff
@@ -27,6 +27,12 @@ def _prepare_merge_moves_distinct_fields(self):
         # merged.
         return super()._prepare_merge_moves_distinct_fields() + ["original_group_id"]
 
+    def _keep_original_group(self):
+        """Keep the original group on the move when creating a joint group."""
+        for move in self:
+            if move.group_id and not move.original_group_id:
```
Comment (rousseldenis) [inline stock_picking_group_by_partner_by_carrier/models/stock_move.py]: Nothing more about printed pickings and state ?

**Review by rousseldenis — APPROVED** [review]
Code review

File: stock_picking_group_by_partner_by_carrier/models/stock_rule.py [hunk 2]
```diff
@@ -2,12 +2,26 @@
...
+        help="Propagate the original group on the created moves. This allows to "
+        "prevent merging pulled moves. This allows to cancel a SO without "
+        "canceling pulled moves from other SO (as we ensure they are not "
+        "merged). If your warehouse is configured to make the delivery in two "
+        "steps, you should not propagate the original group on the picking if your"
+        "pick picking type is also configured to group pickings. In this way, "
+        "the generated moves for the same product but for different sales orders "
+        "will be merged into one unique move. If you propagate the original group "
+        "on the picking, the moves will not be merged and you will have one move "
+        "per sale order.",
```
Comment (jbaudoux) [inline stock_picking_group_by_partner_by_carrier/models/stock_rule.py]: I would not talk about configuring merging on non outgoing moves as this should not be possible. The option is only visible on outgoing picking types. We should add a constrain to make it more explicit.
```suggestion
        "merged). You likely need this option if you don't propagate the merged procurement group to the pulled moves.",
```

File: stock_picking_group_by_partner_by_carrier/tests/test_grouping.py [hunk 3]
```diff
@@ -267,6 +277,73 @@ def test_delivery_multi_step_group_pick(self):
         self.assertEqual(len(picks), 1)
         self.assertFalse(so1.picking_ids - so2.picking_ids)
 
+    def test_delivery_multi_step_group_out_pick_op_merged(self):
```
Comment (jbaudoux) [inline stock_picking_group_by_partner_by_carrier/tests/test_grouping.py]: This should not be possible

**Review by jbaudoux — CHANGES_REQUESTED** [review]
This overall looks good to me

**Comment by lmignon** [comment]
Back to draft. Compatibility issue with the stock_available_to_promise_release.

File: stock_picking_group_by_partner_by_carrier/tests/test_grouping.py (hunk 3 above)
Comment (jbaudoux) [reply stock_picking_group_by_partner_by_carrier/tests/test_grouping.py]: I would drop this test that does not make sense as you should not be able to enable grouping on non outgoing picking type

**Comment by jbaudoux** [comment]
@lmignon Can you drop changes to generated files?

**Comment by lmignon** [comment]
@jbaudoux tests removed and rebase done

**Pull request closed without merging**